src/file.jl

```diff
@@ -154,7 +101,64 @@ tbl = CSV.File(file) |> SQLite.load!(db, "sqlite_table")
 
 $KEYWORD_DOCS
 """
-function File(source;
+struct File <: AbstractVector{Row}
+    name::String
+    names::Vector{Symbol}
+    types::Vector{Type}
+    rows::Int64
+    cols::Int64
+    columns::Vector{Column}
+    lookup::Dict{Symbol, Column}
+end
+
+getname(f::File) = getfield(f, :name)
+getnames(f::File) = getfield(f, :names)
+gettypes(f::File) = getfield(f, :types)
+getrows(f::File) = getfield(f, :rows)
+getcols(f::File) = getfield(f, :cols)
+getcolumns(f::File) = getfield(f, :columns)
+getlookup(f::File) = getfield(f, :lookup)
+getcolumn(f::File, col::Int) = getfield(f, :columns)[col]
+getcolumn(f::File, col::Symbol) = getfield(f, :lookup)[col]
+
+function Base.show(io::IO, f::File)
+    println(io, "CSV.File(\"$(getname(f))\"):")
+    println(io, "Size: $(getrows(f)) x $(getcols(f))")
+    show(io, Tables.schema(f))
+end
+
+Base.IndexStyle(::Type{File}) = Base.IndexLinear()
+Base.eltype(::File) = Row
+Base.size(f::File) = (getrows(f),)
+
+Tables.isrowtable(::Type{File}) = true
+Tables.columnaccess(::Type{File}) = true
+Tables.schema(f::File)  = Tables.Schema(getnames(f), gettypes(f))
+Tables.columns(f::File) = f
+Tables.columnnames(f::File) = getnames(f)
+Base.propertynames(f::File) = getnames(f)
+
+function Base.getproperty(f::File, col::Symbol)
+    lookup = getfield(f, :lookup)
+    return haskey(lookup, col) ? lookup[col].column : getfield(f, col)
+end
+
+function Base.getindex(f::File, col::Symbol)
+    lookup = getfield(f, :lookup)
+    return haskey(lookup, col) ? lookup[col].column : getfield(f, col)
+end
+
+Base.getindex(f::File, col::String) = getindex(f, Symbol(col))
+
+Tables.getcolumn(f::File, nm::Symbol) = getcolumn(f, nm).column
+Tables.getcolumn(f::File, i::Int) = getcolumn(f, i).column
+
+Base.@propagate_inbounds function Base.getindex(f::File, row::Int)
+    @boundscheck checkbounds(f, row)
+    return Row(getnames(f), getcolumns(f), getlookup(f), row)
+end
+
+function File(source::Union{Vector{UInt8}, IO, Cmd, AbstractString, AbstractPath};
```

Comment:
This seems weirdly specific and non-specific...? Maybe you could just special case the byte vector?